src/pdaaal/engine/UntypedPAutomaton.cpp

```diff
@@ -0,0 +1,541 @@
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    void UntypedPAutomaton::_post_star() {
+        // This is an implementation of Algorithm 2 (figure 3.4) in:
+        // Schwoon, Stefan. Model-checking pushdown systems. 2002. PhD Thesis. Technische Universität München.
+        // http://www.lsv.fr/Publis/PAPERS/PDF/schwoon-phd02.pdf (page 48)
+
+        std::vector<std::unique_ptr<temp_edge_t>> edges;
+        std::stack<temp_edge_t *> trans;
+        std::vector<temp_edge_t *> rel;
+        auto dummy = std::make_unique<temp_edge_t>();
+        auto & pda_states = pda().states();
+        auto n_pda_states = pda_states.size();
+        auto insert_edge = [&edges, &dummy, &trans, &rel, this](size_t from, uint32_t label, size_t to,
+                                                                const trace_t *trace,
+                                                                bool direct_to_rel = false) {
+            (*dummy) = temp_edge_t(from, label, to);
+            auto lb = std::lower_bound(std::begin(edges), std::end(edges), dummy,
+                                       [](auto &a, auto &b) -> bool { return *a < *b; });
+            if (lb == std::end(edges) || *(*lb) != *dummy) { // New edge is not already in edges (rel U trans).
+                lb = edges.insert(lb, std::make_unique<temp_edge_t>(from, label, to));
+                if (direct_to_rel) {
+                    rel.push_back(lb->get());
+                } else {
+                    trans.push(lb->get());
+                }
+                if (trace != nullptr) { // Don't add existing edges
+                    if (label == std::numeric_limits<uint32_t>::max()) {
+                        this->add_epsilon_edge(from, to, trace);
+                    } else {
+                        this->add_edge(from, to, label, trace);
+                    }
+                }
+            }
+        };
+
+        // trans := ->_0 intersect (P x Gamma x Q)  (line 1)
+        // rel := ->_0 \ trans (line 2)
+        for (auto &from : this->states()) {
+            for (auto &edge : from->_edges) {
+                assert(!edge.has_epsilon()); // PostStar algorithm assumes no epsilon transitions in the NFA.
+                for (auto &label : edge._labels) {
+                    insert_edge(from->_id, label._label, edge._to->_id, nullptr, from->_id >= n_pda_states);
+                }
+            }
+        }
+
+        // for <p, y> -> <p', y1 y2> do  (line 3)
+        //   Q' U= {q_p'y1}              (line 4)
+        std::unordered_map<std::pair<size_t, uint32_t>, size_t, boost::hash<std::pair<size_t, uint32_t>>> q_prime{};
+        for (auto &state : pda_states) {
+            for (auto &rule : state._rules) {
+                if (rule._operation == PDA::PUSH) {
+                    auto pair = std::make_pair(rule._to, rule._op_label);
+                    if (q_prime.find(pair) == q_prime.end()) {
```

Comment:
Normally you would do the "hashmap.find()" into a separate variable. It can be used as hint for the insertion.
Also, ptries would probably do well here.

Reply:
For unordered_map you can't really get a useful hint like you can with lower_bound for ordered containers. Here the 'hint' would always be std::end(q_prime), which would not be useful. 

ptries would be nice to tryout.